resolve twingly/twingly-search-api-java#9

xSAVIKx opened the pull request:
added betamax functional tests

walro commented on src/test/resources/com/twingly/search/client/tapes/search_without_valid_API_key.yaml:
```diff
@@ -0,0 +1,24 @@
+      Connection: keep-alive
+      Content-Length: '0'
+      Host: api.twingly.com
+      User-Agent: Twingly Search Java Client/null
```
Happened to notice `/null` here, is there some problem with reading the version from the property file?

xSAVIKx replied:
Hm. I've tried several solutions, so have only checked what's in response only for the first ones (everything was OK there). I'll check what could be wrong.

walro commented:
I guess you noticed the trouble with Java 7 `java.lang.UnsupportedClassVersionError` for the betamax test class. I took a look at Betamax and it seems like they with Java 7 so I guess the problem is something else.

walro commented:
Otherwise I think it looks good. How tricky would it be to incorporate this in the Spock/Groovy test suite? :) I understand it might not be worth it, but it would be cool if all tests were using the same test style, if possible.

xSAVIKx commented:
Ahh, yeah, I've forgotten about java7.
I've already written some notes about java7 to betamax team lead [issue 175](https://github.com/betamaxteam/betamax/issues/175#issuecomment-195680647).
I think that the problem could be in published by betamax Jar-file, I think it's compiled using java8 compiller.

walro commented:
> I think that the problem could be in published by betamax Jar-file, I think it's compiled using java8 compiller.

Ah, that makes sense.

xSAVIKx commented:
I've rewritten test with groovy and Spock and also changed slightly VERSION constant, so now it will be 100% loaded after version property is gathered from version.properties.

walro commented:
> I've rewritten test with groovy and Spock and also changed slightly VERSION constant, so now it will be 100% loaded after version property is gathered from version.properties.

Nice, but it looks like a new error was introduced for Java8. Did it run fine locally for you?

xSAVIKx commented:
@walro I think it's some Travis-ci bug, cause on my branch  tests passed for java 8 (as locally they are also OK).

xSAVIKx commented:
I think that branch could be only merged into your repo, when betamax will release a new snapshot. (https://github.com/betamaxteam/betamax/issues/186)

xSAVIKx commented:
Again some weird things happened.
After changing betamax to current snapshot my branch is "green" now... But PR build is broken again. For now have no ideas what could be wrong.

dentarg commented:
@xSAVIKx we are now using the latest snapshot of betamax, and that snapshot now (since 8 hours ago :) ) changes with every commit to the betamax repo

the only real change that have been made to betamax, that I can spot, since we tried and jdk8 worked, is https://github.com/betamaxteam/betamax/pull/180, could that affect what we are doing here? cc @cowboygneox

https://github.com/betamaxteam/betamax/issues/179

xSAVIKx commented:
@dentarg I'd understand, if my fork is failed also, but, you see, my fork was successfully built. And local build is also OK, so I cannot understand what's wrong.
Probably, while travis is trying to build PR it break smth related to self-signed certificates, that are used to handle https calls. It's the only assumption I can do now.

dentarg commented:
Ah that's strange... for the record, here is your green build: https://travis-ci.org/xSAVIKx/twingly-search-api-java/builds/115814335. I restarted our failing build: https://travis-ci.org/twingly/twingly-search-api-java/builds/115814365

cowboygneox commented:
@dentarg The change in https://github.com/betamaxteam/betamax/pull/180 will only affect the default configuration, which will be to use the proxy. In most examples and usages, you won't use the default, and thus should have no impact. If you're not configuring the proxy, you're unlikely using Betamax effectively.

dentarg commented:
This is really strange, https://travis-ci.org/twingly/twingly-search-api-java/builds/115814365 fails but not
https://travis-ci.org/xSAVIKx/twingly-search-api-java/builds/115814335, but both should be the same code. I'm going to check with Travis CI support before merging.

`gradlew check` works locally for me (`1.7.0_45-b18`), so I don't think we have a real problem.

thank you @xSAVIKx for the work here and thank you @cowboygneox for reviving betamax, I will be on the look out for the next relase :)

dentarg commented:
@xSAVIKx when I ran locally, I noticed that `littleproxy_cert` and `littleproxy_keystore.jks` was created in the project root, perhaps we should add those to `.gitignore`?

dentarg commented:
I think I know the problem with Travis, it doesn't seem to run the latest commit... some sync issue

dentarg commented:
> I think I know the problem with Travis, it doesn't seem to run the latest commit... some sync issue

No...

dentarg commented:
I have now emailed travis ci support

dentarg commented:
I also did the merge with master locally and ran the tests, worked fine

cowboygneox commented:
@dentarg You should ignore those certs; they are generated as necessary for SSL support. If you commit them, they will likely expire some time in the future causing messy failures. I'll have to add that to the documentation on Betamax.

xSAVIKx commented:
> @xSAVIKx when I ran locally, I noticed that littleproxy_cert and littleproxy_keystore.jks was created in the project root, perhaps we should add those to .gitignore?

@dentarg , as @cowboygneox already mentioned - these are generated certificates to support SSL, so yes, I think I should add them to .gitignore.

walro commented on build.gradle:
```diff
@@ -18,13 +18,16 @@ ext {
     cglibVersion = "3.2.0"
     objenesisVersion = "2.2"
     jaxbVersion = "2.2.11"
+    betamaxVersion = "2.0.0-alpha-1"
```
What is `betamaxVersion` used for? Seems to me that only `betamaxSnapshotVersion` is used.

dentarg replied:
@walro I think it's left in so we can easily switch back when a proper release of betamax has been made

walro replied:
Yeah, I thought it was something like that. I'd like it removed anyway, but we can do it ourselves.

xSAVIKx replied:
@walro Yeap it was just for simplification of future dependency version change.

walro commented:
@xSAVIKx do you mind trying to restart the tests for your repository? We're trying to figure out the Travis problem. We noticed that you used a slightly older version of Betamax (even though it should not matter).

xSAVIKx commented:
@walro I've just restarted build, so here are the results: [build \#57](https://travis-ci.org/xSAVIKx/twingly-search-api-java/builds/115858471)

walro commented:
> @walro I've just restarted build, so here are the results: build #57

Thanks, the mystery remains unsolved :smile:

dentarg commented:
Just compiling some information

From https://github.com/twingly/twingly-search-api-java/pull/20

"continuous-integration/travis-ci/pr — The Travis CI build failed"
"continuous-integration/travis-ci/push — The Travis CI build passed"

The "Details" link for the failed build takes me to https://travis-ci.org/twingly/twingly-search-api-java/builds/116189066

the Travis page links to "Commit ee5b640" which is https://github.com/twingly/twingly-search-api-java/commit/ee5b640155611e55ebe3b5614e7ceead492a760c (merge commit created by GitHub?)

Take "Details" link for the passed build takes me to https://travis-ci.org/twingly/twingly-search-api-java/builds/116188596

the Travis page links to "Commit 528c51c" which is https://github.com/twingly/twingly-search-api-java/commit/528c51c689dc5b2c5a514562c0b18baa30df6d0a

Raw logs for the build job for oraclejdk8

failed build https://s3.amazonaws.com/archive.travis-ci.org/jobs/116189071/log.txt
passed build https://s3.amazonaws.com/archive.travis-ci.org/jobs/116188602/log.txt

dentarg commented:
> the Travis page links to "Commit ee5b640" which is ee5b640 (merge commit created by GitHub?)

I'm going to play around with https://github.com/twingly/twingly-search-api-java/commit/ee5b640155611e55ebe3b5614e7ceead492a760c.patch

dentarg commented:
> I'm going to play around with https://github.com/twingly/twingly-search-api-java/commit/ee5b640155611e55ebe3b5614e7ceead492a760c.patch

On my iMac:

````
~/twingly/twingly-search-api-java master
$ curl -o the-merge-commit.patch https://github.com/twingly/twingly-search-api-java/commit/ee5b640155611e55ebe3b5614e7ceead492a760c.patch
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 43606    0 43606    0     0  29626      0 --:--:--  0:00:01 --:--:-- 29623

~/twingly/twingly-search-api-java master*
$ git apply --check the-merge-commit.patch

~/twingly/twingly-search-api-java master*
$ git apply the-merge-commit.patch

~/twingly/twingly-search-api-java master*
$ git st
On branch master
Your branch is up-to-date with 'origin/master'.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   .gitignore
	modified:   build.gradle
	modified:   src/main/java/com/twingly/search/Constants.java

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	src/test/groovy/com/twingly/search/client/UrlConnectionClientFunctionSpockTest.groovy
	src/test/resources/com/twingly/search/client/tapes/
	the-merge-commit.patch

no changes added to commit (use "git add" and/or "git commit -a")

~/twingly/twingly-search-api-java master*
$ gradlew check

...

BUILD SUCCESSFUL

$ java -version
java version "1.7.0_45"
Java(TM) SE Runtime Environment (build 1.7.0_45-b18)
Java HotSpot(TM) 64-Bit Server VM (build 24.45-b08, mixed mode)
````

dentarg commented:
> failed build https://s3.amazonaws.com/archive.travis-ci.org/jobs/116189071/log.txt
> passed build https://s3.amazonaws.com/archive.travis-ci.org/jobs/116188602/log.txt

`git diff` at https://gist.github.com/anonymous/333a63e95a6f78215f5c

dentarg commented:
for "continuous-integration/travis-ci/push" we get `WARNING: Playing back from tape ...`
for "continuous-integration/travis-ci/pr" we dont' get that

Pretty sure the problem is that betamax isn't using the tape for "continuous-integration/travis-ci/pr", but as to why, I have no idea...

xSAVIKx commented:
@dentarg betamax tries to use tape, but, I think, smth happened while request was intercepted, cause in PR log, there's INFO:
```
INFO: loaded tape with 1 recorded interactions from file /home/travis/build/twingly/twingly-search-api-java/src/test/resources/com/twingly/search/client/tapes/search_for_spotify.yaml...
----------------------------------------------------
INFO: loaded tape with 1 recorded interactions from file /home/travis/build/twingly/twingly-search-api-java/src/test/resources/com/twingly/search/client/tapes/search_without_valid_API_key.yaml...
```

dentarg commented:
@xSAVIKx thanks for pointing that out, it's helpful

Old but similar issue: https://github.com/betamaxteam/betamax/issues/134 which says

>When the tests is run, it shows:

>        INFO: loaded tape with 1 recorded interactions from file piwik_success.yaml...

> But the tape is not used in tests. The service makes normal HTTP call.

Sounds a bit like what we are experiencing...

dentarg commented:
Also similar https://github.com/betamaxteam/betamax/issues/117

dentarg commented:
The response from Travis CI support:

>We have investigated this on our side and we are unfortunately a bit lost about what is at play here. It seems like the Betamax filters aren't installed in the PR build. Unfortunately, we don't see how and why this can happen.

>I was wondering if it would be possible to add more logs in the Betamax code or even in LittleProxy (where the filters should be installed I believe) to try to understand what's going on. What do you think?

My reply was (which contains some info I haven't shared in this thread yet)

>Not exactly the logs you are asking for, but I made the tests run with --debug --full-stacktrace in my own fork:

>https://github.com/dentarg/twingly-search-api-java/pull/3

>jdk8 fail    https://travis-ci.org/dentarg/twingly-search-api-java/jobs/116498255
jdk8 pass https://travis-ci.org/dentarg/twingly-search-api-java/jobs/116498139

>I will investigate if we can make Betamax and LittleProxy log anything.

>Any ideas on what could be causing this? Anything different in the Travis environment for pr vs push?

That was 2 days ago, no further correspondence since then have taken place.

dentarg commented:
@cowboygneox sorry to drag you into this again, but do you mind taking a closer look? seems the issue here is similar to some old betamax issues, https://github.com/betamaxteam/betamax/issues/134 and https://github.com/betamaxteam/betamax/issues/117. I think the description "It seems like the Betamax filters aren't installed in the PR build" is accurate, but as to why, I have no clue. Any ideas on what logging could be added our Betamax code, or LittleProxy?

dentarg commented:
Hmm... https://github.com/betamaxteam/betamax/pull/193 also fails, and so did https://github.com/betamaxteam/betamax/pull/192 until a race condition was fixed. Related?

dentarg commented:
This branch is also interesting https://github.com/strmer15/betamax/commits/feature/start_once_2 (https://github.com/betamaxteam/betamax/pull/192 is about https://github.com/strmer15/betamax/commits/feature/start_once)

cowboygneox commented:
@dentarg It's no fun waking up to an email in the morning that your library is causing others issues :( I'm looking into it right now.

cowboygneox commented:
@dentarg Ugh, more SSL issues. These seem to be coming up more frequently now that Java enforces it more. I've actually been struggling myself to get this working (though I have it working) but communicating the steps that I used may be just as challenging. Let me keep looking at your config.

cowboygneox commented:
@dentarg (and others observing this): is there any chance you can join the [Betamax Gitter](https://gitter.im/betamaxteam/betamax) so that we can quick round up on this?

dentarg commented:
@cowboygneox sure, I can join

dentarg commented:
Latest from Travis CI

>I had a look at the new log and unfortunately, I could not tell why the Betamax filters don't seem to be installed in the PR build. We are really stumped about how this could be happening and we would really like to understand why it does.

>I can give you more details about the differences between a PR build and a push build:

>1) a PR build runs the merge commit instead of the "branch" commit, so there might be more commits than what shows up in the branch. It's not the case here since it's up-to-date with master.

>2) some `TRAVIS_…` environment variables change, mainly `TRAVIS_PULL_REQUEST` and the commit range one but as far as I can tell, the build doesn't use any of those env vars. Here's a diff I've generated for the environment variables in each build: https://www.diffchecker.com/ovutx8ah.

>3) encrypted secure env vars are removed if the PR comes from a fork but I don't think there are any secure env vars here, am I right?

>Hence, my only other suggestion would be, as I mentioned before, to try instrumenting either or both the Betamax and LittleProxy libraries to better understand why the filters aren't set properly in the PR build. I'm sorry I don't have any easier solution at this time.

>Please let me know if this is something you can/want to try and if you need help doing it. We would be happy to assist you doing so.

My reply

>Correct, no encrypted secure env vars are in use.

>I'm in contact with the current maintainer of Betamax, to debug the problem. My understanding is that he thinks the filters/proxy are used, but there are some problem around SSL certificates (in the pr build), some race condition that are triggered(?)

>If you have any ideas, we are both in the Betamax Gitter at https://gitter.im/betamaxteam/betamax

cowboygneox commented:
@dentarg I think I solved your issues in the latest 2.0.0-SNAPSHOT of Betamax. Can you try pulling it and see if it works?

dentarg commented:
@cowboygneox seems like it stills fails, using https://oss.sonatype.org/content/repositories/snapshots/software/betamax/betamax-junit/2.0.0-SNAPSHOT/betamax-junit-2.0.0-20160415.010203-17.pom, build log at https://travis-ci.org/twingly/twingly-search-api-java/jobs/115858478

Haven't looked into it more than that and I'm afraid I wont be able until some time next week

cowboygneox commented:
@dentarg I submitted another SNAPSHOT, and then I also forked your work and ran tests. Your tests passed fine on my run: https://travis-ci.org/cowboygneox/twingly-search-api-java/builds/123307325. Will you kick yours off again?

dentarg commented:
@cowboygneox Running now https://travis-ci.org/twingly/twingly-search-api-java

cowboygneox commented:
@dentarg Looks like it failed again. I have absolutely no idea why your branch is failing and mine is consistently passing; all I did was fork your branch and ignore some files. We'll probably have to ask travis-ci to compare.

cowboygneox commented:
@dentarg Maybe I need to rebase against master. I'll try that next.

cowboygneox commented:
@dentarg Ok, so I made my own PR #21 and it failed just like this one does, so at least that's consistent. When I run in my own fork, the tests pass, but when we run in a PR, tests fail. I sent an email to travis-ci to have them point out any configuration differences between the types.

walro commented:
It would be nice to finally get this in, @twingly/dev WDYT of the following:

Since the problems we are seeing are likely due to the combination of betamax and https (see #21) we could make the https part configurable for now, defaulting to `true`. When we record, and run, the Betamax tests we "drop down" to http. Once Betamax can handle https we'll revert the changes.

cowboygneox commented:
@walro I am confident that disabling use of `https` will solve your issues.

Also, VCR doesn't fire up a proxy server, it just stubs out the http_client you're using underneath. It doesn't do a true MITM proxy like we're trying to do. We're trying to support more HTTP clients than we have in the past, and the (seemingly) most appropriate way is just to fire up another server to hit.

cowboygneox commented:
@walro Betamax 2.0 official is out. It fixes this issue, but it will require you to change how you run builds on Travis. Please see the documentation on Betamax's readme and let me know if you have any questions.

walro commented:
> @walro Betamax 2.0 official is out. It fixes this issue, but it will require you to change how you run builds on Travis. Please see the documentation on Betamax's readme and let me know if you have any questions.

Thanks, on vacation right now. But once back at the office I'll take a look (unless one of my co-workers beats me to it :-))

dentarg commented:
I'm also on vacation, but I may take a look at it tonight anyways :-)

> On 7 juli 2016, at 19:02, Robin Wallin <notifications@github.com> wrote:
> 
> @walro Betamax 2.0 official is out. It fixes this issue, but it will require you to change how you run builds on Travis. Please see the documentation on Betamax's readme and let me know if you have any questions.
> 
> Thanks, on vacation right now. But once back at the office I'll take a look (unless one of my co-workers beats me to it :-))
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

dentarg commented:
Closed in favor of https://github.com/twingly/twingly-search-api-java/pull/26.
